### DISCUSSION TITLE
termux support

### DISCUSSION THREAD

**Discussion by stsp** [dosemu2/dosemu2#2332] [General]
Hi!

After almost a month of a hard work,
I've finally got the more-or-less working
termux build of dosemu2!
Its even built on CI.
Unfortunately I can only test it under
waydroid/x86_64. Is anyone willing
to test it on some real hardware?

Atrifacts of the first working build are here:
https://github.com/stsp/termux-packages/actions/runs/12507729143/artifacts/2364131044
No deployment yet... So you'll need
to unpack the archive, remove the
`*-cross` debs and install the rest with
`dpkg -i ./*.deb`.
Would be cool if someone can test it!

**Comment by stsp** [comment]
aarch64 link is this one:
https://github.com/stsp/termux-packages/actions/runs/12507729143/artifacts/2364132412
Previous was x86_64, which is
already tested on waydroid.
Though I suspect I need to find
a deployment before people
agree to test anything...

**Reply by andrewbird** [reply]
Since you mention aarch64, I wonder if you're thinking I might test your termux thingy. Unfortunately I've given the raspberry pi5 away so I no longer have access to one of those. BTW the last thing I did was try to build the ia16 debian packages on it, but got stuck in a mutual build dependency between gcc-ia16(-bootstrap?) and newlib. Perhaps that worked on x86 because of existing ia16 packages, but as far as I know there are none for arm64.

**Reply by stsp** [reply]
> I wonder if you're thinking I might test your termux thingy.

Not necessarily you: from you it
would still require installing waydroid,
which is not a pleasant experience, to
say the least.
Its very likely someone else has an
android nettop or a phone. For such
person it would be much easier to
test. But now I've found a few dosemu-
related requests on termux and tur
github trackers, so maybe I can reach
them instead. In fact, people have already
tried to package dosemu2 and even
uploaded some results, eg:
https://github.com/termux-user-repository/tur/pull/1146
But I am sceptical this actually worked,
otherwise why have I spent the full month
on the same thing. :)

> Perhaps that worked on x86 because of existing ia16 packages, but as far as I know there are none for arm64.

What do you mean?
IIRC my PPA has it built for all arches.
Why couldn't you just get it from there?
Also, how about starting to use dj64dev
for new tests? It has demos and lots of
docs. I even documented how to build
ncurses and run its tests, and it is very
trivial to do.

**Reply by stsp** [reply]
I wonder if it makes sence to create
a youtube video about dj64dev. I am
sure people have not anticipated
building ncurses on linux into elf,
then run such elf under DOS...
This is something no one have neither
seen, nor thought is even possible. :)

**Reply by andrewbird** [reply]
> IIRC my PPA has it built for all arches.

RaspiOS is a Debian base not Ubuntu, if you remember I had to build all the non standard dosemu2 build packages for raspberry pi5. 

> I am sure people have not anticipated building ncurses on linux into elf, then run such elf under DOS

Me included! What use case to you see for such binaries?

**Reply by stsp** [reply]
> RaspiOS is a Debian base not Ubuntu

Ok but at least you could steal
the build recipes from the deb pkg.
There is another interesting option
I was thinking about recently, which
is to detach fdpp from freedos and
wrap it around libi86, similar to what
I did for dj64dev. Unfortunately, in
case of libi86 it will require porting
the full thing, whereas for dj64dev I
was able to skip the most dirty parts,
such as far-pointer meta-compiler. So
I won't be doing that alone - its only
needed for freedos cross-compilation,
but freedos ppl are usually not interested
in anything at all.

> Me included!

You don't have to wait for the video,
as you can just follow the "building
something real" chapter of this doc:
https://github.com/stsp/dj64dev
Its really very simple.

> What use case to you see for such binaries?

It would be good to have things like
Gopherus pre-packaged with dosemu2
in an elf form. As this is what you get
w/o sufficiently crippling the build system,
which would be needed to build into .exe.
As you can see from the aforementioned
doc, I build this DOSish elf w/o crippling
the ncurses build system at all, its pristine!
So all I want is a simple cross-compilation.
There is no other task an ELF for DOS
can solve, at least to my knowledge.
And supporting it outside of comcom64
is also not a pleasant task.

**Reply by stuaxo** [reply]
> Since you mention aarch64, I wonder if you're thinking I might test your termux thingy. Unfortunately I've given the raspberry pi5 away so I no longer have access to one of those. BTW the last thing I did was try to build the ia16 debian packages on it, but got stuck in a mutual build dependency between gcc-ia16(-bootstrap?) and newlib. Perhaps that worked on x86 because of existing ia16 packages, but as far as I know there are none for arm64.

Re: raspberry pis: I have a bunch that support aarch64, maybe I could stick one on somewhere you can get access.

I have a 4 not doing anything for instance.

**Reply by stsp** [reply]
Would be cool, but an extra work
for you is to set up waydriod+termux+ssh
there, rather than just ssh.

**Comment by andrewbird** [comment]
> Ok but at least you could steal
the build recipes from the deb pkg.

That's what I was trying to do. But trying to build newlib it said I needed gcc(something) and so I tried to build gcc and it said it needed newlib. Anyway I'm a bit vague here as I no longer have the system, perhaps I'll get another at some point.

**Comment by stsp** [comment]
Termux build finally deployed to tur.

**Reply by stuaxo** [reply]
I added the [tur](https://github.com/termux-user-repository/tur) repo and installed it.

When I run `dosemu` it complains about:

```
~ $ dosemu
CANNOT LINK EXECUTABLE "/data/data/com.termux/files/usr/libexec/dosemu2/dosemu2.bin": library "libandroid-posix-semaphore.so" not found: needed by main executable
```

Looks like it might be a matter of making sure it's linked, e.g.:
https://github.com/termux/termux-packages/issues/21898#issuecomment-2424728446

**Reply by stsp** [reply]
Could you please try the package
(just dosemu2 package alone) from
this build:
https://github.com/stsp/tur/actions/runs/12641252733?pr=1
At the bottom download the needed
artifact and extract the needed deb
from the archive.

**Comment by stuaxo** [comment]
I had to install libelf as it was complaining.

Then I got to here -

CONF: emulated CPU forced down to real CPU: 386      ERROR: term: stderr still on tty, closing            Terminated with signal 31

I remembered I need to run X, when I tried to specify SDL.

I need to do this again with a cable so I can control it via scrcpy from my laptop, it's pretty painful doing it from the phone.

**Comment by stuaxo** [comment]
I got further and installed xfce, when I run dosemu in the terminal there, on the phone I get:

??? ?  ~ dosemu
CONF: emulated CPU forced down to real CPU: 386
ERROR: ladspa: failed to load filter.so
ERROR: libao: unable to open output device
ERROR: soundfonts not found
madvise(): Invalid argument
madvise(): Invalid argument
Can't create debugger pipes, dosdebug not available
Terminated with signal 31

**Comment by stsp** [comment]
I fixed most of deps problems on tur.
You probably now need to manually
uninstall my build and reinstall the
tur build to get it updated.

**Reply by stuaxo** [reply]
Excellent, that actually boots to C:\> and I can navigate around, I should probably sort out some software to run :)

**Reply by stsp** [reply]
Try `insfdusr`

**Reply by stuaxo** [reply]
`insfdusr` fails because `tqdm` (progress bar library for python) isn't installed.  On ubuntu that package is `python3-tqdm`.

I've requested a package here -
https://github.com/termux-user-repository/tur/issues/1442

I did get working installing `tqdm` from pip, but I think having a native tur package would be best - let's see if it gets added (tqdm is pretty popular).

**Comment by stsp** [comment]
I wonder if now we can create an f-droid
app using termux as a dependency or
maybe using termux-libs for cross-compile...
But I am out of bandwidth on any termux
tasks, so maybe someone else can take
a look at if its even possible or not.
I installed f-droid under waydroid/qemu-aarch64
and it just crashes on start, whereas everything
else works, so f-droid is likely buggy.

**Comment by stsp** [comment]
That was the wrong thing to do.
Latest insfdusr installs tqdm, you
were not up-to-date (or I did
something wrong)

Понедельник, 13 января 2025 г получено от Stuart Axon:

**Reply by stuaxo** [reply]
Pretty sure I have the latest one (from tur at least) - unless I need the latest build from here?

Is tqdm vendored into the dosemu package ?

`insfdusr` can't find it and I can't importing it from a python shell also fails.

I do python dev and there are quite a few packages that use tqdm, so it would be good for more than just dosemu.

**Reply by stsp** [reply]
https://github.com/termux-user-repository/tur/commit/99fb4650cc3fd813b2951ec583f81a2466af4c64
This was supposed to install tqdm as a dep.
Check the version of instfd package.

**Reply by stuaxo** [reply]
OK, managed to get this after uninstalling + updating the repo reinstalling: I'm not sure which stage I missed before.

I mostly develop in python + having a system package depend on one from pip can cause issues later on, and generally things get messy (also know this from maintaining shoebot for a long time, where we have depended on packages in pip but not on the system).

tqdm is all sorts of places apart from here, I think I'll still leave the request open for `python3-tqdm`: at work, we're using it in Jupyter notebooks - and I've used it and it's predecessor in various places over the last decade +.

**Comment by stsp** [comment]
Hmm, its possible that postinst script is not activated on update, as it is supposed to be executed just once.
If you want a feature request, you can request an extension on termux to generate such postinst script automatically, or handle python deps by some other means - automatically.
Manually producing postinst is silly and unreliable.

Вторник, 14 января 2025 г получено от Stuart Axon:

**Reply by stuaxo** [reply]
I don't know...  the last time I installed it and it worked, I did see that get triggered.

The time it worked, I additionally had done dist-upgrade (which didn't seem to do anything), and uninstalled instfdusr - which also installed `dosemu`.

I reckon on a fresh system it may well work...  now I think of it, I should have been trying it on the Astro slide I have sitting in a box, it has a keyboard so may be slightly less hellish to use termux on.

**Reply by stsp** [reply]
Yep, I have no idea why dosemu
would be useful on phone. Astro
slide may be a better fit.
I am quite sceptical about this
whole termux thing, but the
fact that it can work there, is
cool by itself.

**Comment by stuaxo** [comment]
![image](https://github.com/user-attachments/assets/6ae93f96-fb50-494f-a1fb-0e56c3b0cce4)

Random "productivity" screen shot.  Almost makes me want to break our turbo pascal and experiment building some touch optimised UI for DOS.

EDIT:  had no idea how large this pic was until I came back to the browser on the computer, have rescalled to 50%

**Reply by stsp** [reply]
Hmm, but its not bad after all!
If not for the fact that DOS software
is completely useless on a phone,
it looks cool. :)

**Reply by stuaxo** [reply]
Pretty useless.

Though QuickBasic, Turbo Pascal and Turbo C++ all had pretty integrated help.

So it could be possible to generate more useless software this way.

It certainly is a weird thing seeing it like this, which is enjoyable.
